File: sydney/Kernel/Common/NullData.cpp
// -*-Mode: C++; tab-width: 4; c-basic-offset: 4;-*-
// vi:set ts=4 sw=4:
//
// NullData.cpp -- NULL データ関連の関数定義
// 
// Copyright (c) 1999, 2001, 2004, 2005, 2006, 2007, 2010, 2023 Ricoh Company, Ltd.
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// 

namespace {
const char srcFile[] = __FILE__;
const char moduleName[] = "Common";
}

#include "SyDefault.h"
#include "SyReinterpretCast.h"

#include "Common/ClassID.h"
#include "Common/NullData.h"
#include "Common/UnicodeString.h"

#include "Exception/BadArgument.h"

#include <iostream>
using namespace std;

_TRMEISTER_USING
_TRMEISTER_COMMON_USING

namespace
{
}

//	FUNCTION public
//	Common::NullData::getInstance -- NullDataのインスタンスを返す
//
//	NOTES
//		NullData のインスタンスはこの関数でしか得られない
//
//	ARGUMENTS
//		なし
//
//	RETURN
//		得られた NullData のインスタンスを格納する領域の先頭アドレス
//
//	EXCEPTIONS
//		なし

// static
const NullData*
NullData::getInstance()
{
	// NullData のインスタンスを新規に確保せずに
	// 即得るための予約されたインスタンス
	static const NullData reserved;

	return &reserved;
}

//	FUNCTION private
//	Common::NullData::NullData -- デフォルトコンストラクタ
//
//	NOTES
//
//	ARGUMENTS
//		なし
//
//	RETURN
//		なし
//
//	EXCEPTIONS
//		なし

NullData::NullData()
	: ScalarData(DataType::Null)
{}

//
//	FUNCTION public
//	Common::NullData::~NullData -- デストラクタ
//
//	NOTES
//	デストラクタ
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	なし
//
//	EXCEPTIONS
//	なし
//
NullData::
~NullData()
{
}

//	FUNCTION public
//	Common::NullData::serialize -- シリアル化
//
//	NOTES
//
//	ARGUMENTS
//	ModArchive& cArchiver_
//		アーカイバー
//	RETURN
//	なし
//
//	EXCEPTIONS

void
NullData::serialize(ModArchive& cArchiver_)
{
	Data::serialize_NotNull(cArchiver_);
}

//
//	FUNCTION public
//	Common::NullData::copy -- コピーする
//
//	NOTES
//	自分自身のコピーを返す。
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	Common::Data::Pointer
//		自分自身のコピー
//
//	EXCEPTIONS

Data::Pointer
NullData::copy() const
{
	return NullData::getInstance();
}

//
//	FUNCTION public
//	Common::NullData::getString -- 文字列でデータを得る
//
//	NOTES
//	文字列でデータを得る
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	ModUnicodeString
//		文字列にしたデータ
//
//	EXCEPTIONS
//	なし
//
ModUnicodeString
NullData::getString() const
{
	return _TRMEISTER_U_STRING("(null)");
}

// FUNCTION public
//	Common::NullData::assign -- 代入を行う
//
// NOTES
//
// ARGUMENTS
//	const Data* r
//	
// RETURN
//	bool
//
// EXCEPTIONS
//		Exception::BadArgument
//			Common::NullData には代入できない

//virtual
bool
NullData::
assign(const Data* r, bool bForAssign_ /* = true */)
{
	_TRMEISTER_THROW0(Exception::BadArgument);
}

// FUNCTION public
//	Common::NullData::equals_NoCast -- 等しいか調べる(キャストなし)
//
// NOTES
//
// ARGUMENTS
//	const Data& r
//	
// RETURN
//	bool
//
// EXCEPTIONS

//virtual
bool
NullData::
equals_NoCast(const Data& r) const
{
	return r.isNull();
}

// FUNCTION public
//	Common::NullData::compareTo_NoCast -- 大小比較を行う(キャストなし)
//
// NOTES
//
// ARGUMENTS
//	const Data& r
//	
// RETURN
//	int
//
// EXCEPTIONS

//virtual
int
NullData::
compareTo_NoCast(const Data& r) const
{
	// Nullのほうが小さいと扱う
	return r.isNull() ? 0 : -1;
}

//	FUNCTION public
//	Common::Data::setNull -- NULL 値にする
//
//	NOTES
//
//	ARGUMENTS
//		bool		v
//			true または指定されないとき
//				NULL 値にする
//			false
//				NULL 値でなくする
//
//	RETURN
//		なし
//
//	EXCEPTIONS
//		Exception::BadArgument
//			Common::NullData は NULL 値でなくせない

void
NullData::setNull(bool v)
{
	if (!v)
		_TRMEISTER_THROW0(Exception::BadArgument);

	// DEFAULTではなくなる
	setDefault(false);
}

//	FUNCTION public
//	Common::NullData::isAbleToDump -- ダンプできるか調べる
//
//	NOTES
//
//	ARGUMENTS
//		なし
//
//	RETURN
//		true
//			ダンプできる
//		false
//			ダンプできない
//
//	EXCEPTIONS

bool
NullData::isAbleToDump() const
{
	return Data::isAbleToDump_NotNull();
}

//	FUNCTION public
//	Common::NullData::isFixedSize -- 常に固定長であるかを得る
//
//	NOTES
//
//	ARGUMENTS
//		なし
//
//	RETURN
//		true
//			固定長である
//		false
//			固定長でない
//
//	EXCEPTIONS
//		なし

bool
NullData::isFixedSize() const
{
	return Data::isFixedSize_NotNull();
}

//	FUNCTION public
//	Common::NullData::getDumpSize -- ダンプサイズを得る
//
//	NOTES
//
//	ARGUMENTS
//		なし
//
//	RETURN
//		得られたダンプサイズ(B 単位)
//
//	EXCEPTIONS

ModSize
NullData::getDumpSize() const
{
	return Data::getDumpSize_NotNull();
}

//	FUNCTION public
//	Common::NullData::dumpValue -- 自分の値を与えられた領域にダンプする
//
//	NOTES
//
//	ARGUMENTS
//		char* 		dst
//			自分の値をダンプする領域の先頭
//
//	RETURN
//		ダンプサイズ(B 単位)
//

ModSize
NullData::dumpValue(char* dst) const
{
	return Data::dumpValue_NotNull(dst);
}

ModSize
NullData::dumpValue(ModSerialIO& cSerialIO_) const
{
	return Data::dumpValue_NotNull(cSerialIO_);
}

//
//	FUNCTION public
//	Common::NullData::getClassID -- クラスIDを得る
//
//	NOTES
//	シリアライズのためのクラスIDを得る。
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	int
//		Common::NullData クラスのクラスID
//
//	EXCEPTIONS
//	なし
//
int
NullData::
getClassID() const
{
	return ClassID::NullDataClass;
}

//
//	FUNCTION public
//	Common::NullData::print -- 値を表示する
//
//	NOTES
//	値を表示する
//
//	ARGUMENTS
//	なし
//
//	RETURN
//	なし
//
//	EXCEPTIONS
//	なし
//
void
NullData::
print() const
{
	cout << "null: " << endl;
}

// FUNCTION public
//	Common::NullData::hashCode -- ハッシュコードを取り出す
//
// NOTES
//
// ARGUMENTS
//	Nothing
//
// RETURN
//	ModSize
//
// EXCEPTIONS

//virtual
ModSize
NullData::
hashCode() const
{
	return 0;
}

//
//	Copyright (c) 1999, 2001, 2004, 2005, 2006, 2007, 2010, 2023 Ricoh Company, Ltd.
//	All rights reserved.
//

File: sydney/Java/JDBC/src/jp/co/ricoh/doquedb/common/NullData.java
// -*-Mode: Java; tab-width: 4; c-basic-offset: 4;-*-
// vi:set ts=4 sw=4:
//
// NullData.java -- NULLをあらわすクラス
//
// Copyright (c) 2002, 2003, 2004, 2023 Ricoh Company, Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

package jp.co.ricoh.doquedb.common;

/**
 * NULLをあらわすクラス
 *
 */
public final class NullData extends ScalarData
	implements Serializable
{
	/** NullDataのインスタンス */
	private static NullData nullData = new NullData();

	/** NullDataのインスタンスを得る */
	public static NullData getInstance() { return nullData; }

	/**
	 * データ型をあらわすクラスを新たに作成する
	 */
	public NullData()
	{
		super(DataType.NULL);
	}

	/**
	 * ストリームから読み込む
	 *
	 * @param input_	入力用のストリーム
	 * @throws java.io.IOException
	 *			入出力関係の例外が発生した
	 * @see Serializable#readObject(InputStream) readObject
	 */
	public void readObject(InputStream input_)
		throws java.io.IOException
	{
	}

	/**
	 * ストリームに書き出す
	 *
	 * @param output_	出力用のストリーム
	 * @throws java.io.IOException
	 *			入出力関係の例外が発生した
	 * @see	Serializable#writeObject(OutputStream) writeObject
	 */
	public void writeObject(OutputStream output_)
		throws java.io.IOException
	{
	}

	/**
	 * データが等しいか比較する
	 *
	 * @return	等しい場合はtrue、それ以外の場合はfalse
	 */
	public boolean equals(Object other_)
	{
		boolean result = false;
		if ((other_ instanceof NullData) == true)
		{
			result = true;
		}
		return result;
	}

	/**
	 * オブジェクトのコピーを作成して返す
	 *
	 * @return	コピーされたオブジェクト
	 */
	public Object clone()
	{
		return NullData.getInstance();
	}

	/**
	 * 文字列に変換する
	 *
	 * @return	文字列
	 */
	public String toString()
	{
		return "(null)";
	}

	/**
	 * {@link ClassID クラスID}を得る
	 *
	 * @return {@link ClassID クラスID}
	 * @see Serializable#getClassID() getClassID
	 */
	public int getClassID()
	{
		return ClassID.NULL_DATA;
	}
}

//
// Copyright (c) 2002, 2003, 2004, 2023 Ricoh Company, Ltd.
// All rights reserved.
//
